### PR TITLE
Borrowing & publishing from non published parts

### PR DESCRIPTION
**Pull request by Coenego**
This is a placeholder / braindump ticket to handle a couple of (edge) cases around borrowing:

**Borrowing from the same module in which you would like to borrow**
In borrowing modal, the current module should be visible, but disabled

**Publishing a part which has series in it which belong to a part which hasn't been published yet**
Admins should be able to borrow series from parts which are not published yet, with the principle that unpublished series from other departments shouldn't hold up timetable creation.

On the student UI, we could do this to deal with this case:

Unpublished series should be displayed, but the Add button disabled. If somebody clicks on Add event button, a tooltip could be displayed with the following copy:

This {{event_type}} is belongs to {{tripos}} / {{part}} which is not published yet.
You will be able to add this to your calendar once it is published.

### CODE REVIEW

**Comment by Coenego** [comment]
Assigning to @bp323 for review

File: shared/gh/js/utils/gh.utils.orgunits.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    var decorateBorrowedSeriesWithParentInfo = exports.decorateBorrowedSeriesWithParentInfo = function(series) {
+        _.each(series, function(serie) {
+
+            // Only decorate borrowed series
+            /* istanbul ignore else */
+            if (serie.borrowedFrom) {
+                _.each(_triposData, function(orgUnitType) {
+
+                    // Add the parent info to the organisational unit, if any
+                    var parent = _.find(orgUnitType, function(orgUnit) { return orgUnit.id === serie.borrowedFrom.ParentId; });
```
Comment (bp323) [inline shared/gh/js/utils/gh.utils.orgunits.js]: Can you split this into multiple lines to be consistent?

File: shared/gh/js/utils/gh.utils.orgunits.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+     * Decorate an organisational unit with its parent info
+     *
+     * @param  {Object}    orgUnit    The organisational unit that needs to be decorated with his parent info
+     */
+    var addParentInfoToOrgUnit = exports.addParentInfoToOrgUnit = function(orgUnit) {
+        if (orgUnit.ParentId) {
+            _.each(_triposData, function(_orgUnitType) {
+
+                // Add the parent info to the organisational unit, if any
+                var parent = _.find(_orgUnitType, function(_orgUnit) { return _orgUnit.id === orgUnit.ParentId; });
```
Comment (bp323) [inline shared/gh/js/utils/gh.utils.orgunits.js]: Can you split this into multiple lines?

File: shared/gh/js/utils/gh.utils.orgunits.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+     */
+    var getTriposStructure = exports.getTriposStructure = function(appId, callback) {
+        if (!_.isFunction(callback)) {
+            throw new Error('An invalid value for callback was provided');
+        } else if (appId && !_.isNumber(appId)) {
+            throw new Error('An invalid value for appId was provided');
+        }
+
+        var core = require('gh.core');
+        /* istanbul ignore next */
```
Comment (bp323) [inline shared/gh/js/utils/gh.utils.orgunits.js]: Can you test this? Should just be a case of leaving out the appId when calling the function and checking that the proper things return.

File: shared/gh/js/utils/gh.utils.orgunits.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                var parent = _.find(_orgUnitType, function(_orgUnit) { return _orgUnit.id === orgUnit.ParentId; });
+                if (parent) {
+                    orgUnit.Parent = parent;
+                }
+            });
+        }
+    };
+
+    /**
+     * Return the tripos structure
```
Comment (bp323) [inline shared/gh/js/utils/gh.utils.orgunits.js]: While we're changing this, can you expand on the JSDocs a bit to include what the returned tripos structure object looks like?

File: shared/gh/js/views/gh.series-borrowed-popover.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    // BINDING //
+    /////////////
+
+    /**
+     * Add handlers to the popover triggers
+     *
+     * @private
+     */
+    var addBinding = function() {
+
```
Comment (bp323) [inline shared/gh/js/views/gh.series-borrowed-popover.js]: Empty line can be removed

File: tests/qunit/tests/js/api.util.js [hunk 6]
```diff
@@ -899,18 +900,134 @@ require(['gh.core', 'gh.api.tests'], function(gh, testAPI) {
...
         // Verify that an error is thrown when no callback was provided
         assert.throws(function() {
-            gh.utils.getTriposStructure();
+            gh.utils.getTriposStructure(null);
         }, 'Verify that an error is thrown when no callback was provided');
 
         // Verify that an error is thrown when an invalid callback was provided
         assert.throws(function() {
-            gh.utils.getTriposStructure('invalid_callback');
+            gh.utils.getTriposStructure(null, 'invalid_callback');
```
Comment (bp323) [inline tests/qunit/tests/js/api.util.js]: I placed a comment above asking to test with an empty appId but looks like you're already doing this. Can you just go ahead and remove the `istanbul ignore` from that line in the utility file?

**Comment by bp323** [comment]
Functional issue:

When hitting 'subscribe to all' in the student view, borrowed series that haven't been published yet look like they've been borrowed as well. Refreshing the page still actually shows that the borrowed, unpublished, series has been borrowed. I'm not sure if this is a UI issue or backend.

**Comment by bp323** [comment]
Looks good, back to @Coenego for some followup

**Comment by Coenego** [comment]
@bp323 It actually is a back-end issue. When you want to 'subscribe to all', we pass in the module ID. The back-end automatically subscribes the user to all the event series in that module, regardless if the original part was published or not. Not much we can do at the moment...

**Comment by Coenego** [comment]
Reassigning to @bp323 for review after follow-up

**Comment by bp323** [comment]
Verified and merged.

**Pull request merged**